mne/preprocessing/artifact_detection.py

```diff
@@ -4,12 +4,113 @@
+
+    Detects data segments containing activity in the frequency range given by
+    ``filter_freq`` whose envelope magnitude exceeds the specified z-score
+    threshold, when summed across channels and divided by ``sqrt(n_channels)``.
+    False-positive transient peaks are prevented by low-pass filtering the
+    resulting z-score time series at 4 Hz. Only operates on a single channel
+    type, if ``ch_type`` is ``None`` it will select the first type in the list
+    ``mag``, ``grad``, ``eeg``.
+    See :footcite:`Muthukumaraswamy2013` for background on choosing
+    ``filter_freq`` and ``threshold``.
+
+    Parameters
+    ----------
+    raw : instance of Raw
+        Data to estimate segments with muscle artifacts.
+    threshold : float
+        The threshold in z-scores for marking segments as containing muscle
+        activity artifacts.
+    ch_type : 'mag' | 'grad' | 'eeg' | None
+        The type of sensors to use. If ``None`` it will take the first type in
+        ``mag``, ``grad``, ``eeg``.
+    min_length_good : float | None
+        The shortest allowed duration of "good data" (in seconds) between
+        adjacent annotations; shorter segments will be incorporated into the
+        surrounding annotations.``None`` is equivalent to ``0``.
+        Default is ``0.1``.
+    filter_freq : array-like, shape (2,)
+        The lower and upper frequencies of the band-pass filter.
+        Default is ``(110, 140)``.
+    %(n_jobs)s
+    %(verbose)s
+
+    Returns
+    -------
+    annot : mne.Annotations
+        Periods with muscle artifacts annotated as BAD_muscle.
+    scores_muscle : array
+        Z-score values averaged across channels for each sample.
+
+    References
+    ----------
+    .. footbibliography::
+    """
+    from scipy.stats import zscore
+    from scipy.ndimage.measurements import label
+
+    raw_copy = raw.copy()
+
+    if ch_type is None:
+        raw_ch_type = raw_copy.get_channel_types()
+        if 'mag' in raw_ch_type:
+            ch_type = 'mag'
+        elif 'grad' in raw_ch_type:
+            ch_type = 'grad'
+        elif 'eeg' in raw_ch_type:
+            ch_type = 'eeg'
+        logger.info('Using %s sensors for muscle artifact detection'
+                    % (ch_type))
+
+    if ch_type in ('mag', 'grad'):
+        raw_copy.pick_types(meg=ch_type, ref_meg=False)
+    elif ch_type == 'eeg':
+        raw_copy.pick_types(meg=False, eeg=True)
+
+    raw_copy.filter(filter_freq[0], filter_freq[1], fir_design='firwin',
+                    pad="reflect_limited", n_jobs=n_jobs)
+    raw_copy.apply_hilbert(envelope=True, n_jobs=n_jobs)
+
+    data = raw_copy.get_data(reject_by_annotation="NaN")
+    nan_mask = ~np.isnan(data[0])
+    sfreq = raw_copy.info['sfreq']
+
+    art_scores = zscore(data[:, nan_mask], axis=1)
+    art_scores = art_scores.sum(axis=0) / np.sqrt(art_scores.shape[0])
+    art_scores = filter_data(art_scores, sfreq, None, 4)
```

Comment:
filtering by default ommits bad annotation segments, here you are filtering across the 'edges' left after removing periods where annotations were - so you can smooth out jumps that arise due to removing parts of the signal. it would be better if this was fixed - but it can be done in a subsequent PR.